File: 816. Ambiguous Coordinates/solution.cpp
class Solution {
public:
    vector<string> gen(string s) {
        int length = s.size();
        if (length == 0 || (length > 1 && s[0] == '0' && s[length-1] == '0')) {
            return {};
        }
        if (s[0] == '0' && length > 1) {
            return {"0." + s.substr(1)};
        }
        if (s[length-1] == '0' || length == 1) {
            return {s};
        }
        vector<string> result = {s};
        for (int i = 1; i < length; ++i) {
            result.push_back(s.substr(0, i) + '.' + s.substr(i));
        }
        return result;
    }  
    
    vector<string> ambiguousCoordinates(string s) {
        int length = s.size();
        vector<string> result;
        
        for (int i = 1; i < length-2; ++i) {
            vector<string> left = gen(s.substr(1, i));
            vector<string> right = gen(s.substr(i + 1, length - 2 - i));
            
            for (auto& a: left) {
                for (auto& b: right) {
                    result.push_back("(" + a + ", " + b + ")");
                }
            }
        }
        
        return result;
    }
};

File: 816. Ambiguous Coordinates/solution.py
class Solution:
    def generate_float(self, s: str):
        if not s or len(s) > 1 and s[0] == s[-1] == '0': 
            return []
        if s[0] == '0':
            if len(s) == 1:
                return[s]
            else:
                return ['0.' + s[1:]]
        if s[-1] == '0':
            return [s]
        floats = [s]
        for idx in range(1, len(s)):
            floats.append(s[:idx] + '.' + s[idx:])
        return floats
    
    def ambiguousCoordinates(self, s: str) -> List[str]:
        result = []
        
        # remove brackets
        s = s[1:-1]
        
        for idx in range(len(s)):
            for left, right in itertools.product(self.generate_float(s[:idx]), 
                                                 self.generate_float(s[idx:])):
                result.append(f"({left}, {right})")
            
        return result
        
